tests/tls_e2e/server_enc/openssl_server.cpp

```diff
@@ -0,0 +1,260 @@
+#include <openenclave/internal/raise.h>
+#include <openenclave/internal/report.h>
+#include <openenclave/internal/syscall/device.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+#include <arpa/inet.h>
+#include <sys/socket.h>
+#include <unistd.h>
+
+#include <openssl/evp.h>
+#include <openssl/ssl.h>
+// clang-format on
+
+#include "tls_e2e_t.h"
+#include "../common/openssl_utility.h"
+
+extern "C"
+{
+    int setup_tls_server(struct tls_control_args* config, char* server_port);
+    int launch_tls_client(
+        struct tls_control_args* config,
+        char* server_name,
+        char* server_port);
+};
+
+struct tls_control_args g_control_config;
+
+int launch_tls_client(
+    struct tls_control_args* config,
+    char* server_name,
+    char* server_port)
+{
+    (void)config;
+    (void)server_name;
+    (void)server_port;
+    return 0;
+}
+
+int create_listener_socket(uint16_t port, int& server_socket)
+{
+    int ret = -1;
+    struct sockaddr_in addr;
+    addr.sin_family = AF_INET;
+    addr.sin_port = htons(port);
+    addr.sin_addr.s_addr = htonl(INADDR_ANY);
+
+    server_socket = socket(AF_INET, SOCK_STREAM, 0);
+    if (server_socket < 0)
+    {
+        OE_TRACE_ERROR(TLS_SERVER "socket creation failed \n");
+        goto exit;
+    }
+
+    if (bind(server_socket, (struct sockaddr*)&addr, sizeof(addr)) < 0)
+    {
+        OE_TRACE_ERROR(TLS_SERVER "Unable to bind socket to the port \n");
+        goto exit;
+    }
+
+    if (listen(server_socket, 20) < 0)
+    {
+        OE_TRACE_ERROR(TLS_SERVER "Unable to open socket for listening \n");
+        goto exit;
+    }
+    ret = 0;
+exit:
+    return ret;
+}
+
+int handle_communication_until_done(
+    int& server_socket_fd,
+    int& client_socket_fd,
+    SSL_CTX*& ssl_server_ctx,
+    SSL*& ssl_session,
+    bool keep_server_up)
+{
+    int ret = -1;
+
+waiting_for_connection_request:
+
+    // reset ssl_session setup and client_socket_fd to prepare for the new TLS
+    // connection
+    close(client_socket_fd);
+    SSL_free(ssl_session);
+    OE_TRACE_INFO(" waiting for client connection \n");
+
+    struct sockaddr_in addr;
+    uint len = sizeof(addr);
+    client_socket_fd = accept(server_socket_fd, (struct sockaddr*)&addr, &len);
+
+    if (client_socket_fd < 0)
+    {
+        OE_TRACE_ERROR("Unable to accept the client request \n");
+        goto exit;
+    }
+
+    // create a new SSL structure for a connection
+    if ((ssl_session = SSL_new(ssl_server_ctx)) == NULL)
+    {
+        OE_TRACE_ERROR(TLS_SERVER
+                       "Unable to create a new SSL connection state object\n");
+        goto exit;
+    }
+
+    SSL_set_fd(ssl_session, client_socket_fd);
+
+    // wait for a TLS/SSL client to initiate a TLS/SSL handshake
+    if (SSL_accept(ssl_session) <= 0)
+    {
+        OE_TRACE_ERROR(" SSL handshake failed \n");
+        ret = -0x3000; // To match with SSL handshake error for mbedtls so that
```

Comment:
This is interesting. Do we check the return value? If so, can you point it out? Instead of matching the Mbed TLS implementation, we can update the test cases.

Reply:
It would return <=0 value on failure. https://www.openssl.org/docs/man1.0.2/man3/SSL_accept.html.
This is where we check the return value on host side. https://github.com/openenclave/openenclave/blob/master/tests/tls_e2e/host/host.cpp#L166

Reply:
Given this code is only defined in this test, I'd suggest update the test case to check the different code in Mbed TLS- and OpenSSL-based implementations.